File: js/sim/patrician.ts
// THE PATRICIAN — a diegetic peace-keeper (Discworld's Lord Vetinari). The Director
// is the agent of drama; the Patrician is its counterweight. Each measured interval
// it scans the town for the most dangerous FEUD (the most mutually-hostile pair of
// townsfolk — often one the Director just lit) and BROKERS a partial truce, quelling
// any latched intra-town hostility before it becomes a killing.
//
// The point is not to remove tension but to MANAGE it: feuds still smoulder, crime
// still simmers, but the city holds — so the world stays interesting without tearing
// itself apart. Belief-only (it nudges standings; ground truth is untouched), no
// gold, fully guarded (never throws on the fixed tick).

import { PATRICIAN } from './simconfig.js';
import { rng } from './rng.js';
import { brandForsworn } from './houses.js';

// `sim`/`ctx` (the owning Simulation + cognition context — wave-2, still .js) and the
// agents (via their belief/standing flags) are typed opaquely on purpose; this layer is
// belief-only and fully guarded, so behaviour is unchanged.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Sim = any;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Ag = any;

export class Patrician {
  sim: Sim;
  _acc: number;
  stats: Record<string, number>;

  constructor(sim: Sim) {
    this.sim = sim;
    this._acc = 0;
    this.stats = { truces: 0, quelled: 0, forswornLeaks: 0 };
  }

  tick(ctx: unknown, dt: number): void {
    try {
      if (!this.sim._spawned) return;   // a real town only (not bare test sub-sims)
      // BETRAYAL-AS-CHOICE leak (every tick, PROMPT): drain any oath-abandonment stamped on a
      // breaker's own-state this cognition tick and leak it to its witnesses BEFORE the breaker
      // wanders off. Cheap (a single O(n) roster pass; short-circuits when nothing is armed).
      this._leakForsworn();
      this._acc += dt;
      if (this._acc < PATRICIAN.tickEvery) return;
      this._acc = 0;
      this._broker();
    } catch { /* never throw on the tick */ }
  }

  // OBSERVER bridge for betrayal-as-choice: motivation.resolveOath stamps `_forswornLeak` on a
  // breaker's own-state (cognition writes only its own flag — the epistemic split holds). Here, in
  // the omniscient society pass, we read the roster to PLANT the deed into nearby witnesses' beliefs
  // (reputation.witnessForsworn — each writes its OWN belief) and, if the breaker is now a habitual
  // oathbreaker, brand it. The marker is consumed once. Belief/name only; ground truth untouched.
  _leakForsworn(): void {
    const agents = this.sim.agents;
    if (!agents || !this.sim.reputation || !this.sim.reputation.witnessForsworn) return;
    for (const a of agents) {
      const mark = a && (a as { _forswornLeak?: { t: number; count: number } })._forswornLeak;
      if (!mark) continue;
      (a as { _forswornLeak?: unknown })._forswornLeak = undefined;   // consume once
      if (!a.alive) continue;
      try {
        const n = this.sim.reputation.witnessForsworn(agents, a, this.sim.time);
        if (n > 0) {
          this.stats.forswornLeaks = (this.stats.forswornLeaks || 0) + 1;
          this._note(`Word spreads that ${a.name} forsook a sworn vow.`);
        }
        brandForsworn(this.sim, a);   // gated on life.forsworn; brands once, never clobbers a held epithet
      } catch { /* never throw on the tick */ }
    }
  }

  // find the single most-mutually-hostile pair of townsfolk and broker a truce.
  _broker(): void {
    const folk = this.sim.agents.filter((a: any) => a.alive && a.autonomous && a.faction === 'townsfolk');
    // a real intra-town FEUD is a latched HOSTILE belief between two townsfolk (what
    // the director's feud / a kin-vendetta leaves behind) — NOT merely a low standing,
    // which a fond pair soured by a spark never crosses. Find the most bitter such pair.
    let pair = null, worst = Infinity;
    for (const A of folk) {
      const all = A.beliefs && A.beliefs.all ? A.beliefs.all() : [];
      for (const b of all) {
        if (!b) continue;
        const B = this.sim.agentsById.get(b.subjectId);
        if (!B || !B.alive || B.faction !== 'townsfolk' || B === A) continue;
        const ba = B.beliefs.get(A.id);
        if (!(b.hostile || (ba && ba.hostile))) continue;   // only an actual feud (latched hostility)
        const sum = (b.standing || 0) + ((ba && ba.standing) || 0);
        if (sum < worst) { worst = sum; pair = [A, B]; }
      }
    }
    if (!pair) return;
    const [A, B] = pair;
    // most interventions are a TRUCE (damp the feud); occasionally the Patrician
    // achieves a lasting RECONCILIATION — former enemies become friends. Peace, not
    // just a pause: the positive counterpoint to all the blood-feuds.
    // CONTESTED / REFUSABLE PEACE: brokering is not omnipotent. A party nursing a fresh, salient
    // BETRAYED/AVENGE wound RESISTS — a forced forgiveness over real blood is no forgiveness. A
    // reconciliation (enemies → friends) is impossible if EITHER side holds such a wound; it
    // degrades to a (still-resisted) truce. A failed brokering that hardens a feud is more
    // interesting than silent neutralization.
    const grudge = Math.max(this._grudgeSalience(A, B), this._grudgeSalience(B, A));
    if (grudge < (PATRICIAN.refuseSalience || 0.6) && rng() < (PATRICIAN.reconcileChance || 0)) {
      this._reconcile(A, B);
    } else {
      const r1 = this._truce(A, B);
      const r2 = this._truce(B, A);
      this.stats.truces++;
      if (r1 || r2) this._note(`The Patrician tried to broker peace between ${A.name} and ${B.name}, but the grudge runs deep.`);
      else this._note(`The Patrician brokered peace between ${A.name} and ${B.name}.`);
    }
  }

  // The peak salience of a fresh BETRAYED/AVENGE-class wound A carries about B (own-memory only —
  // the epistemic split: we read the TARGET's own autobiography to gauge how forgivable this is).
  // Scans all rings; a hostile-feud grudge memory keyed on B (withId/byId) at high salience is the
  // "I will not be made to forgive this" signal. Bounded, guarded, returns 0 when there's none.
  _grudgeSalience(A: Ag, B: Ag): number {
    try {
      const m = A && A.memory;
      if (!m || !B) return 0;
      const rings = [m.stm, m.mtm, m.ltm];
      let best = 0;
      const GRUDGE = new Set(['betrayed', 'assaulted', 'forsworn', 'witnessed_death', 'grief']);
      for (const r of rings) {
        if (!r || !r.items) continue;
        for (const e of r.items()) {
          if (!e || (e.valence || 0) >= 0) continue;
          if (e.withId !== B.id && e.byId !== B.id) continue;
          if (!GRUDGE.has(e.kind)) continue;
          if ((e.salience || 0) > best) best = e.salience || 0;
        }
      }
      return best;
    } catch { return 0; }
  }

  // Pull A's standing toward B back up toward neutral, defuse a latched hostility — UNLESS A holds a
  // salient grudge about B, in which case the nudge is throttled by (1 - grudge): a deep wound barely
  // moves, and the hostility is NOT quelled. When the wound is severe, the forced peace can BACKFIRE —
  // A resents being made to forgive: the feud hardens a touch and A files a `forced_peace` resentment
  // memory (and, rarely, sours on the Patrician's whole project — recorded, never a hidden truth flip).
  // Returns true if A resisted (so the caller can narrate a hardened feud rather than a clean peace).
  _truce(A: Ag, B: Ag): boolean {
    const b = A.beliefs && A.beliefs.get ? A.beliefs.get(B.id) : null;
    if (!b) return false;
    const grudge = this._grudgeSalience(A, B);
    const throttle = Math.max(0, 1 - grudge);              // 1 = no grudge (full nudge); 0 = unforgivable
    const resisted = grudge >= (PATRICIAN.resistSalience || 0.5);
    if (resisted && rng() < (PATRICIAN.hardenChance || 0)) {
      // BACKFIRE: forcing the issue hardens the feud instead of damping it.
      b.standing = Math.max(-1, (b.standing || 0) - (PATRICIAN.hardenAmount || 0.08));
      if (A.memory) { try { A.memory.record({ t: this.sim.time, kind: 'forced_peace', withId: B.id, valence: -1, salience: Math.min(0.85, 0.45 + grudge * 0.4) }); } catch { /* */ } }
      this.stats.hardened = (this.stats.hardened || 0) + 1;
      return true;
    }
    b.standing = Math.min(0.1, (b.standing || 0) + (PATRICIAN.brokerAmount || 0) * throttle);
    if (PATRICIAN.quellHostile && b.hostile && !resisted) { b.hostile = false; this.stats.quelled++; }
    return resisted;
  }

  // a LASTING peace: mutual standing turns warmly POSITIVE (friendship, so it won't
  // simply re-sour), hostility is cleared, and each records a reconciliation bond.
  // If the two belong to different Houses, it's a peace between their lines — a
  // saga-worthy beat (the marriage-alliance / feud's-end trope).
  _reconcile(A: Ag, B: Ag): void {
    for (const [x, y] of [[A, B], [B, A]]) {
      const b = x.beliefs && x.beliefs.get ? x.beliefs.get(y.id) : null;
      if (b) { b.standing = Math.max(b.standing || 0, PATRICIAN.reconcileStanding || 0.4); b.hostile = false; }
      if (x.memory) { try { x.memory.record({ t: this.sim.time, kind: 'bond', withId: y.id, rel: 'reconciled', valence: 1, salience: 0.7 }); } catch { /* */ } }
    }
    this.stats.reconciliations = (this.stats.reconciliations || 0) + 1;
    const houses = A.house && B.house && A.house !== B.house;
    const text = houses
      ? `Houses ${A.house} and ${B.house} have set aside their feud and made peace.`
      : `${A.name} and ${B.name}, long at odds, have reconciled.`;
    this._note(text, 'legend');   // peace endures in the saga
  }

  _note(text: string, kind?: string): void { try { if (this.sim.chronicle && this.sim.chronicle.note) this.sim.chronicle.note(kind || 'patrician', kind === 'legend' ? -14 : -12, text); } catch { /* */ } }
}

File: rust/sim-core/src/systems/patrician.rs
//! FAN-OUT UNIT: the Patrician (a diegetic peace-keeper — Discworld's Lord Vetinari). Ports the
//! SPIRIT of `js/sim/patrician.ts`. The Director is the agent of drama; the Patrician is its
//! counterweight. On a slow throttle it scans the town for the single most DANGEROUS feud (the most
//! mutually-hostile pair of living townsfolk — often one the Director just lit) and BROKERS a partial
//! truce, quelling latched intra-town hostility before it becomes a killing.
//!
//! The point is not to remove tension but to MANAGE it: feuds still smoulder, but the city holds, so
//! the world stays interesting without tearing itself apart. Belief-only (it nudges standings; ground
//! truth is untouched) — the OBSERVER layer (it narrates by SEEDING beliefs, never by driving an
//! agent's decision; the epistemic split holds). Runs SERIALLY in the society phase.
//!
//! WHAT THIS IMPLEMENTS (throttled `PATRICIAN_EVERY`):
//! - `_broker` — find the most mutually-soured LATCHED-HOSTILE living-townsfolk pair (scan beliefs in
//!   id order; minimize standing-sum; lowest-id tie-break), then:
//!   - `_reconcile` — if NEITHER side carries a deep grudge (an `Assaulted` memory of the other) and a
//!     `sim_rng` roll passes: a lasting peace. Both standings warm to a warmly-POSITIVE bond and
//!     hostility is cleared. A `BEAT_RECONCILE` chronicle beat marks the saga-worthy feud's-end.
//!   - `_truce` — otherwise a partial damp BOTH ways: warm each standing toward neutral and un-latch
//!     hostility — UNLESS that side holds a deep grudge (`_grudgeSalience`), in which case the forced
//!     peace RESISTS (the nudge is withheld and the hostility is NOT quelled — a forced forgiveness
//!     over real blood is no forgiveness). A resisted brokering leaves the feud smouldering.
//!
//! NOT PORTED (noted): `_leakForsworn` (the broken-vow reputation leak) needs a forsworn/oath-broken
//! marker on the agent that does not exist in the Wave substrate yet — SKIPPED by design; only the
//! truce-brokering core is ported.
//!
//! Determinism: SERIAL society phase ⇒ trivially M-invariant (M=1 ≡ M=N). The reconcile roll draws
//! from `world.sim_rng` (never per-entity rng). No HashMap iteration / float reduce in the scan — the
//! pair search is an integer-keyed id-order scan with an (i32 standing-sum, ids) tie-break. No gold,
//! no spawn ⇒ conservation is untouched.

use crate::components::{Beat, EpisodeKind, Faction};
use crate::world::World;

/// How often the Patrician wakes to broker peace (sparse — the counterweight to the director's drama).
const PATRICIAN_EVERY: u32 = 150;

/// A reconcile is only attempted when NEITHER side carries a deep grudge; this is its `sim_rng` roll
/// (the rest of the time a brokering degrades to a partial truce). Mirrors `PATRICIAN.reconcileChance`.
const RECONCILE_CHANCE: f32 = 0.30;

/// Belief-standing nudge toward neutral applied by a (non-resisted) truce, in the i16 quantization
/// (−32768..32767 ≡ −1..1). Mirrors `PATRICIAN.brokerAmount`. A truce damps but does not befriend.
const BROKER_AMOUNT: i16 = 9_000;

/// The warm POSITIVE standing both sides are lifted to by a lasting reconciliation (so it won't simply
/// re-sour). Mirrors `PATRICIAN.reconcileStanding` (~+0.4). `warm_belief` also un-latches hostility.
const RECONCILE_STANDING: i16 = 13_000;

/// BeatKind for a brokered reconciliation (the feud's-end / saga beat; chronicle-local).
const BEAT_RECONCILE: u8 = 22;

pub fn tick(world: &mut World) {
    // Throttle: only broker on the interval boundary (and never at tick 0).
    if world.tick == 0 || world.tick % PATRICIAN_EVERY != 0 {
        return;
    }
    broker(world);
}

/// Find the single most-mutually-hostile pair of living townsfolk and broker a truce (or, rarely, a
/// lasting reconciliation). A real intra-town FEUD is a LATCHED hostile belief between two townsfolk
/// (what the director's feud / a kin-vendetta leaves behind) — NOT merely a low standing, which a fond
/// pair soured by a spark never crosses. We pick the most bitter such pair (minimum standing-sum;
/// lowest-(a,b) ids break ties for determinism).
fn broker(world: &mut World) {
    let (a, b) = match worst_feud(world) {
        Some(p) => p,
        None => return,
    };

    // A reconciliation (enemies → friends) is impossible if EITHER side carries a deep, salient grudge
    // (an `Assaulted` memory of the other) — a forced forgiveness over real blood is no forgiveness; it
    // degrades to a (still-resisted) truce. `grudge_a`/`grudge_b` are the per-side wounds.
    let grudge_a = world.memory[a as usize].has(EpisodeKind::Assaulted, b);
    let grudge_b = world.memory[b as usize].has(EpisodeKind::Assaulted, a);
    let deep_grudge = grudge_a || grudge_b;

    if !deep_grudge && world.sim_rng.next_f32() < RECONCILE_CHANCE {
        reconcile(world, a, b);
    } else {
        // partial truce BOTH ways; each side resists iff IT holds the deep grudge.
        truce(world, a, b, grudge_a);
        truce(world, b, a, grudge_b);
    }
}

/// A LASTING peace: mutual standing turns warmly POSITIVE (friendship, so it won't simply re-sour) and
/// hostility is cleared (`warm_to` un-latches it). Both directions are seeded. A saga-worthy beat is
/// logged (the marriage-alliance / feud's-end trope).
fn reconcile(world: &mut World, a: u32, b: u32) {
    warm_to(world, a as usize, b, RECONCILE_STANDING);
    warm_to(world, b as usize, a, RECONCILE_STANDING);
    world.chronicle.push(Beat {
        t: world.tick,
        kind: BEAT_RECONCILE,
        subject: a,
        magnitude: b as i32,
    });
}

/// Pull `observer`'s standing toward `subject` back up toward neutral and un-latch the hostility —
/// UNLESS `resisted` (the observer holds a deep grudge about the subject), in which case the nudge is
/// WITHHELD entirely and the latched hostility is left in place: a forced peace over real blood is no
/// forgiveness, and the feud smoulders on (more interesting than silent neutralization).
fn truce(world: &mut World, observer: u32, subject: u32, resisted: bool) {
    if resisted {
        return;
    }
    // a non-resisted truce warms toward neutral; `warm_belief` (amt > 0) clears the hostile flag.
    world.warm_belief(observer as usize, subject, BROKER_AMOUNT);
}

/// Warm `observer`'s standing toward `subject` to at LEAST `target` (a reconciliation lifts to a fixed
/// warm floor, not a relative nudge). Seeds the belief if absent, raises a lower standing to `target`,
/// and un-latches hostility. Leaves an already-warmer bond untouched.
fn warm_to(world: &mut World, observer: usize, subject: u32, target: i16) {
    if let Some(ix) = world.ensure_belief(observer, subject) {
        let b = &mut world.beliefs[observer].bodies[ix];
        if b.standing < target {
            b.standing = target;
        }
        b.flags &= !0x01; // un-latch hostile (a lasting peace)
        if b.confidence < 26_000 {
            b.confidence = 26_000;
        }
    }
}

/// The most mutually-hostile pair of living townsfolk, or `None`. A pair qualifies iff `a` holds a
/// belief about `b` AND at least one of the two beliefs is LATCHED hostile (an actual feud, not a mere
/// low standing). The "bitterness" key is `standing(a→b) + standing(b→a)` (lower = worse); ties break
/// on the lowest `(a, b)` ids. Deterministic id-order scan — no float reduce, no HashMap iteration.
fn worst_feud(world: &World) -> Option<(u32, u32)> {
    let n = world.n;
    let mut best: Option<(i32, u32, u32)> = None; // (standing-sum, a, b)
    for a in 0..n {
        if !is_town(world, a) {
            continue;
        }
        for body in world.facts[a].views() {
            let subj = body.subject;
            let b = subj as usize;
            if b >= n || b == a || !is_town(world, b) {
                continue;
            }
            // the reciprocal belief (b about a), if held.
            let ba = world.facts[b].view(a as u32);
            let a_host = body.flags & 0x01 != 0;
            let b_host = ba.map(|x| x.flags & 0x01 != 0).unwrap_or(false);
            if !(a_host || b_host) {
                continue; // only an actual feud (latched hostility on either side)
            }
            let sum = body.standing as i32 + ba.map(|x| x.standing as i32).unwrap_or(0);
            // minimize the standing-sum; lowest (a, b) ids break ties (deterministic).
            let cand = (sum, a as u32, subj);
            match best {
                Some(cur) if cur <= cand => {}
                _ => best = Some(cand),
            }
        }
    }
    best.map(|(_, a, b)| (a, b))
}

/// A living townsperson (the only faction the Patrician brokers between — an intra-town peace-keeper).
#[inline]
fn is_town(world: &World, i: usize) -> bool {
    world.alive[i] && world.faction[i] == Faction::Townsfolk as u8
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::components::{Episode, Faction};
    use crate::world::World;

    /// Seed a one-sided LATCHED-hostile belief: `observer` holds `standing` toward `subject`, hostile.
    fn seed_hostile(w: &mut World, observer: usize, subject: u32, standing: i16) {
        w.sour_belief(observer, subject, 0, true);
        if let Some(ix) = w.beliefs[observer].find(subject) {
            w.beliefs[observer].bodies[ix].standing = standing;
        }
    }

    fn all_townsfolk(w: &mut World) {
        for i in 0..w.n {
            w.faction[i] = Faction::Townsfolk as u8;
            w.alive[i] = true;
            w.beliefs[i].len = 0;
            w.memory[i].len = 0;
        }
    }

    /// A mutually-hostile pair (no deep grudge) gets brokered toward neutral / reconciled: both
    /// standings rise above where they started and neither side stays latched hostile.
    #[test]
    fn hostile_pair_gets_reconciled() {
        let mut w = World::spawn(0xA7C1A, 6);
        all_townsfolk(&mut w);
        // 0 and 1 hold a mutual deep-negative latched-hostile feud (no Assaulted memory ⇒ forgivable).
        seed_hostile(&mut w, 0, 1, -20_000);
        seed_hostile(&mut w, 1, 0, -20_000);

        // run brokering directly (deterministic) — repeat so even a truce path lifts standings.
        w.mirror_beliefs_to_facts();
        for _ in 0..8 {
            broker(&mut w);
        }

        let s01 = w.beliefs[0].find(1).map(|ix| w.beliefs[0].bodies[ix].standing).unwrap();
        let s10 = w.beliefs[1].find(0).map(|ix| w.beliefs[1].bodies[ix].standing).unwrap();
        assert!(s01 > -20_000, "0's standing toward 1 should warm toward neutral, got {s01}");
        assert!(s10 > -20_000, "1's standing toward 0 should warm toward neutral, got {s10}");
        let h01 = w.beliefs[0].find(1).map(|ix| w.beliefs[0].bodies[ix].flags & 0x01).unwrap();
        let h10 = w.beliefs[1].find(0).map(|ix| w.beliefs[1].bodies[ix].flags & 0x01).unwrap();
        assert_eq!(h01, 0, "0 should no longer be latched-hostile to 1");
        assert_eq!(h10, 0, "1 should no longer be latched-hostile to 0");
    }

    /// A deep grudge (an `Assaulted` memory) RESISTS the truce: that side's standing/hostility is left
    /// untouched (a forced forgiveness over real blood is no forgiveness), and no reconcile fires.
    #[test]
    fn deep_grudge_resists_brokering() {
        let mut w = World::spawn(0x9C0DE, 6);
        all_townsfolk(&mut w);
        seed_hostile(&mut w, 0, 1, -20_000);
        seed_hostile(&mut w, 1, 0, -20_000);
        // agent 0 was ASSAULTED by 1 ⇒ 0 will not forgive (its side resists; reconcile is blocked).
        w.memory[0].record(Episode {
            kind: EpisodeKind::Assaulted as u8,
            place: 0,
            valence: -1,
            _pad: 0,
            with: 1,
            t: 0,
            salience: 60_000,
            _pad2: 0,
        });

        for _ in 0..8 {
            broker(&mut w);
        }

        // 0 holds the grudge ⇒ its standing toward 1 is unmoved and still latched hostile.
        let s01 = w.beliefs[0].find(1).map(|ix| w.beliefs[0].bodies[ix].standing).unwrap();
        let h01 = w.beliefs[0].find(1).map(|ix| w.beliefs[0].bodies[ix].flags & 0x01).unwrap();
        assert_eq!(s01, -20_000, "the wounded side resists the truce (standing unchanged)");
        assert_eq!(h01, 0x01, "the wounded side stays latched-hostile (no forced forgiveness)");
        // and no reconcile beat was ever logged (a deep grudge blocks the lasting peace).
        assert!(
            !w.chronicle.iter().any(|bt| bt.kind == BEAT_RECONCILE),
            "a deep grudge must block a reconciliation"
        );
    }

    /// No latched hostility ⇒ no feud ⇒ the Patrician does nothing (a merely-low standing is not a
    /// feud; a fond pair soured by a spark never crosses into a brokerable feud).
    #[test]
    fn low_standing_without_hostility_is_not_a_feud() {
        let mut w = World::spawn(0xCA1F, 6);
        all_townsfolk(&mut w);
        // mutual low standing but NOT latched hostile.
        w.sour_belief(0, 1, 5_000, false);
        w.sour_belief(1, 0, 5_000, false);
        assert!(worst_feud(&w).is_none(), "low standing without hostility is not a feud");
    }

    /// Determinism: the full sim (incl. the patrician society pass) is order-independent across rayon
    /// pool sizes (M=1 ≡ M=N), proven via the world golden hash.
    #[test]
    fn society_patrician_m_invariant() {
        use crate::hash::world_hash;
        use crate::{in_pool, run_sim};
        let h1 = in_pool(1, || world_hash(&run_sim(0x9A11, 400, 80)));
        let h4 = in_pool(4, || world_hash(&run_sim(0x9A11, 400, 80)));
        assert_eq!(h1, h4, "patrician society pass must be M-invariant");
    }
}
